RecoTracker/TkDetLayers/src/Phase2EndcapSubDisk.cc

```diff
@@ -0,0 +1,309 @@
+    theComps.push_back(rings[i]);
+    fillRingPars(i);
+    theBasicComps.insert(
+        theBasicComps.end(), (*rings[i]).basicComponents().begin(), (*rings[i]).basicComponents().end());
+  }
+
+  LogDebug("TkDetLayers") << "==== DEBUG Phase2EndcapSubDisk =====";
+  LogDebug("TkDetLayers") << "r,zed pos  , thickness, innerR, outerR: " << this->position().perp() << " , "
+                          << this->position().z() << " , " << this->specificSurface().bounds().thickness() << " , "
+                          << this->specificSurface().innerRadius() << " , " << this->specificSurface().outerRadius();
+}
+
+BoundDisk* Phase2EndcapSubDisk::computeDisk(const vector<const Phase2EndcapSingleRing*>& rings) const {
+  float theRmin = rings.front()->specificSurface().innerRadius();
+  float theRmax = rings.front()->specificSurface().outerRadius();
+  float theZmin = rings.front()->position().z() - rings.front()->surface().bounds().thickness() / 2;
+  float theZmax = rings.front()->position().z() + rings.front()->surface().bounds().thickness() / 2;
+
+  for (vector<const Phase2EndcapSingleRing*>::const_iterator i = rings.begin(); i != rings.end(); i++) {
+    float rmin = (**i).specificSurface().innerRadius();
+    float rmax = (**i).specificSurface().outerRadius();
+    float zmin = (**i).position().z() - (**i).surface().bounds().thickness() / 2.;
+    float zmax = (**i).position().z() + (**i).surface().bounds().thickness() / 2.;
+    theRmin = min(theRmin, rmin);
+    theRmax = max(theRmax, rmax);
+    theZmin = min(theZmin, zmin);
+    theZmax = max(theZmax, zmax);
+  }
+
+  float zPos = (theZmax + theZmin) / 2.;
+  PositionType pos(0., 0., zPos);
+  RotationType rot;
+
+  return new BoundDisk(pos, rot, new SimpleDiskBounds(theRmin, theRmax, theZmin - zPos, theZmax - zPos));
+}
+
+Phase2EndcapSubDisk::~Phase2EndcapSubDisk() {
+  for (auto c : theComps)
+    delete c;
+
+  delete theComponents.load();
+}
+
+void Phase2EndcapSubDisk::groupedCompatibleDetsV(const TrajectoryStateOnSurface& startingState,
+                                                 const Propagator& prop,
+                                                 const MeasurementEstimator& est,
+                                                 std::vector<DetGroup>& result) const {
+  std::array<int, 3> const& ringIndices = ringIndicesByCrossingProximity(startingState, prop);
+  if (ringIndices[0] == -1 || ringIndices[1] == -1 || ringIndices[2] == -1) {
+    edm::LogError("TkDetLayers") << "TkRingedForwardLayer::groupedCompatibleDets : error in CrossingProximity";
+    return;
+  }
+
+  //order is: rings in front = 0; rings in back = 1
+  //rings should be already ordered in r
+  //if the layer has 1 ring, it does not matter
+  //FIXME: to be optimized once the geometry is stable
+  std::vector<int> ringOrder(theRingSize);
+  std::fill(ringOrder.begin(), ringOrder.end(), 1);
+  if (theRingSize > 1) {
+    if (fabs(theComps[0]->position().z()) < fabs(theComps[1]->position().z())) {
```

Comment:
as a random example of the problems of duplicating code, this violates CMS code rules...

Reply:
@jpata What aspect of it is in violation of the rules? (I assume it's something from the technical coding rules but it's not obvious to me what :-) )

Reply:
I think `std::abs` is to be preferred to `fabs`

Reply:
right. not that this particular example is the most critical bug ever, but it's just confusing to me what is the CMS / trkpog expectation in the case when old code is being copied, how do we want to treat this in the review (potentially also in the old code then).